src/emcee.jl

```diff
@@ -0,0 +1,227 @@
+struct Ensemble{D} <: MHSampler
+    n_walkers::Int
+    proposal::D
+end
+
+struct Walker{T<:Union{Vector,Real,NamedTuple},L<:Real}
+    params::T
+    lp::L
+end
+
+Walker(t::Transition) = Walker(t.params, t.lp)
+Walker(model::DensityModel, params) = Walker(params, logdensity(model, params))
+
+# Define the first step! function, which is called at the 
+# beginning of sampling. Return the initial parameter used
+# to define the sampler.
+function AbstractMCMC.step!(
+    rng::Random.AbstractRNG,
+    model::DensityModel,
+    spl::Ensemble,
+    N::Integer,
+    ::Nothing;
+    init_params = nothing,
+    kwargs...,
+)
+    if init_params === nothing
+        return propose(spl, model)
+    else
+        return Transition(model, init_params)
+    end
+end
+
+# Define the other step functions. Returns a Transition containing
+# either a new proposal (if accepted) or the previous proposal 
+# (if not accepted).
+function AbstractMCMC.step!(
+    rng::Random.AbstractRNG,
+    model::DensityModel,
+    spl::Ensemble,
+    ::Integer,
+    params_prev;
+    kwargs...,
+)
+    # Generate a new proposal. Accept/reject happens at proposal level.
+    return propose(spl, model, params_prev)
+end
+
+#
+# Initial proposal
+# 
+function propose(spl::Ensemble, model::DensityModel)
+    # Make the first proposal with a static draw from the prior.
+    static_prop = StaticProposal(spl.proposal.proposal)
+    mh_spl = MetropolisHastings(static_prop)
+    return map(x -> Walker(propose(mh_spl, model)), 1:spl.n_walkers)
+end
+
+
+
+#
+# Every other proposal
+# 
+function propose(spl::Ensemble, model::DensityModel, walkers::Vector{W}) where {W<:Walker}
+    new_walkers = Vector{W}(undef, spl.n_walkers)
+    interval = 1:spl.n_walkers
+
+    for i in interval
+        walker = walkers[i]
+        other_walker = rand(walkers[interval.!=i])
+        new_walkers[i] = move(spl, model, walker, other_walker)
+    end
+
+    return new_walkers
+end
+
+
+#####################################
+# Basic stretch move implementation #
+#####################################
+struct StretchProposal{P, F<:AbstractFloat} <: Proposal{P}
+    proposal :: P
+    stretch_length::F
+end
+
+StretchProposal(p) = StretchProposal(p, 2.0)
+
+function move(
+    # spl::Ensemble,
+    spl::Ensemble{<:StretchProposal},
+    model::DensityModel,
+    walker::Walker,
+    other_walker::Walker,
+)
+    # Calculate intermediate values
+    proposal = spl.proposal
+    n = length(walker.params)
+    a = proposal.stretch_length
+    z = ((a - 1.0) * rand() + 1.0)^2.0 / a
```

Comment:
Better not hard code floating point numbers (also the compiler optimizes `^2` but I'm not sure if that's possible for `^2.0` as well) and to use the RNG:
```suggestion
    z = ((a - 1) * rand(rng) + 1)^2 / a
```
It would also be nice to not use `log(rand(rng))` but `-randexp(rng)` instead, but I don't see how the expression here could be transformed into that form.